src/common/ServerPushSetupBanner.js

```diff
@@ -0,0 +1,93 @@
+/* @flow strict-local */
+
+import React from 'react';
+import type { Node } from 'react';
+import subWeeks from 'date-fns/subWeeks';
+
+import ZulipBanner from './ZulipBanner';
+import { useSelector, useGlobalSelector, useDispatch } from '../react-redux';
+import { getIdentity, getAccount } from '../account/accountsSelectors';
+import { getIsAdmin, getRealm, getGlobalSettings } from '../directSelectors';
+import { dismissServerPushSetupNotice } from '../account/accountActions';
+import { openLinkWithUserPreference } from '../utils/openLink';
+import { useRoute } from '../react-navigation';
+
+type Props = $ReadOnly<{||}>;
+
+/**
+ * A "nag banner" saying the server hasn't enabled push notifications, if so
+ *
+ * If on the notification-settings screen, it is always visible unless the
+ * server is set up for push notifications.
+ *
+ * Otherwise, it offers a dismiss button. If this notice is dismissed, it
+ * sleeps for two weeks, then reappears if the server hasn't gotten set up
+ * for push notifications in that time. ("This notice" means the currently
+ * applicable notice. If the server does get setup for push notifications,
+ * then gets un-setup, a new notice will apply.)
+ */
+export default function PushNotifsSetupBanner(props: Props): Node {
+  const dispatch = useDispatch();
+
+  // If we can detect that this banner is on the notification-settings
+  // screen, it shouldn't be dismissable.
+  const route = useRoute();
+  const isDismissable = route.name !== 'notifications';
```

Comment:
Hmm, I don't like having this reach upward like this to have its behavior depend on where it finds itself -- it feels like a sort of action-at-a-distance that's going to be a surprise for someone making changes in the future.

It's good that the jsdoc clearly mentions this, and that'd be an essential thing to do to mitigate the confusingness if we did go down this path.

But how about instead making this flag just be a boolean prop? Then `NotificationsScreen` will set it, and `HomeScreen` will leave it out, and it'll be clear from looking at the two call sites why they behave differently, just like normal.

Reply:
Sure, that's reasonable! 🙂

I think I was thinking of a possible future where `<ServerPushSetupBanner />`, `<ServerCompatBanner />`, etc., are called by something like a `BannerManager` component, instead of called directly by our screen components. That `BannerManager` might somehow take care of the [don't-show-two-banners-at-once requirement](https://material.io/components/banners), like by popping up a second applicable banner just after you dismiss the first.

In that case, it might be cumbersome to plumb a prop down to `ServerPushSetupBanner` from the screen component.

But we aren't using something like a `BannerManager` component, and if we were, we might still be concerned about action-at-a-distance.

Reply:
Yeah. We can cross that bridge when we come to it -- we'll see if there's a reasonable way to pass the information through directly, and if not then we can do this or something else indirect.